src/python/WMCore/Services/WMStats/WMStatsPycurl.py

```diff
@@ -0,0 +1,119 @@
+"""
+This module provides some functions to create concurrent HTTP requests
+for CouchDB, as an alternative to the sequential modules available under:
+* WMCore/Services/WMStats/WMStatsReader.getTaskJobSummaryByRequest
+* WMCore/Services/WMStats/WMStatsReader.jobDetailByTasks
+
+Documentation for the parameters supported in views query can be found at:
+https://docs.couchdb.org/en/stable/api/ddoc/views.html#get--db-_design-ddoc-_view-view
+"""
+import json
+
+from urllib.parse import urljoin, urlencode
+from Utils.CertTools import ckey, cert
+from WMCore.Services.pycurl_manager import getdata as multi_getdata
+
+
+def getTaskJobSummaryByRequestPycurl(rowsSummary, sampleSize, serviceOpts):
+    """
+    Pycurl-based implementation of WMStatsReader.getTaskJobSummaryByRequest
+    :param rowsSummary: a dictionary with rows from CouchDB
+    :param sampleSize: integer with number of documents to retrieve
+    :param serviceOpts: dictionary with CouchDB options (url, db name and couchapp name)
+    :return: a dictionary with job detail
+    """
+    paramQueries = []
+    for row in rowsSummary['rows']:
+        thisQuery = {"startkey": [], "endkey": [], "numOfError": 0}
+        # row["key"] = ['workflow', 'task', 'jobstatus', 'exitCode', 'site']
+        thisQuery["startkey"] = row["key"][:4]
+        if row["key"][4]:
+            thisQuery["startkey"].append(row["key"][4])  # site
+
+        thisQuery["endkey"] = []
+        thisQuery["endkey"].extend(thisQuery["startkey"])
+        thisQuery["endkey"].append({})
+        # append amount of errors matching the 5 keys above
+        thisQuery["numOfError"] = row["value"]
+        # add query to the pool
+        paramQueries.append(thisQuery)
+
+    return jobDetailByTasksPycurl(paramQueries, sampleSize, serviceOpts)
+
+
+def jobDetailByTasksPycurl(queries, limit, serviceOpts):
+    """
+    Pycurl-based implementation of WMStatsReader.jobDetailByTasks.
+    In short, for each tuple of errors, it returns "limit" number of documents.
+    :param queries: list of CouchDB query parameters
+    :param limit: number of documents to retrieve from CouchDB
+    :param serviceOpts: dictionary with CouchDB options (url, db name and couchapp name)
+    :return: dictionary with job detail information
```

Comment:
please provide structure of returned dictionary as it is very nested and complex.

Reply:
My second comment is that I would like to know if combined `jobInfoDoc ` is really required here and would it be beneficial to convert this function to a generator to reduce memory footprint. But yielding a generator will affect other code re-factoring.

Reply:
Alan, I put this comment in exact place where clarification is required. I'm talking about structure of return `jobInfoDoc` dictionary and not structure of couchdb calls which you provided. As far as I can read what `jobInfoDoc` looks like it has the following structure:
```
{ workflow_name : {
           task: {
               job_status: {
                    exit_code: { 
                         site: {.....}
                    }
               }      
           }
      }
}
```
In my opinion it would be useful to have it in docstring with full structure (including parts of site dictionary).